### PR TITLE
When you make a client request, if the server is down try to connect first.

### PR DESCRIPTION
**Pull request by RongxinZhang**
The MicroService Clients attempt to connect on an interval with a bit of an exponential backoff. When you restart a service it can take a while before its re-connects.
 - [x] if a method call is made, attempt an immediate reconnect before responding with an error

##### Additional Info

when you start up a microserver client but no server exists, or the server has been disconnected. It will automatically try to reconnect, but at an interval that grows exponentially. 
###### steps
- Say we start the client, it reaches max interval or reconnect which is 20000 millis
- We then start up the server right after a reconect attemp
- We then make a call to the server using the client

###### Result
- We get an error saying server not connected ( which is correct BUT)

###### Expected
- We could attempt a reconnect first, and once connected we could call the method

### CODE REVIEW

**Comment by RongxinZhang** [comment]
@MadManMathew @EnFinlay please review. 

Steps to send call:
- Check if connected
- If not connected -> connect
- Wait for 1 second and check if connected,
- If not connected, log err and callback with error
- else send message

The next time that it sends a call:
- If connected -> check again and send call

**Comment by RongxinZhang** [comment]
Wercker is failing, and we will move this to travis.

**Comment by MadManMathew** [comment]
:+1:

**Comment by thalesfsp** [comment]
@RongxinZhang Bad to merge?

**Comment by RongxinZhang** [comment]
@thalesfsp Just merged in master, should be good after.

**Comment by RongxinZhang** [comment]
One more thumb fro eric and it should be good.

**Comment by thalesfsp** [comment]
:+1:

File: test/connection.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+var Server = require('../lib').Server;
+var Client = require('../lib').Client;
+
+var server;
+var client;
+
+before(function(done){
+  server = new Server('test', 5000);
+  client = new Client('test', 5000)
+  done()
```
Comment (EnFinlay) [inline test/connection.js]: Don't need the `done` since it's not async

File: lib/client.js [hunk 2]
```diff
@@ -90,24 +90,40 @@ MicroServiceClient.prototype.sendCall = function sendCall(method){
...
+      checkAndSendCall();
+    }, 1000);
+  } else {
+    checkAndSendCall();
+  }
 
+  // check and send call
+  function checkAndSendCall() {
+    // check if connected to server
+    if (self.isConnected() === undefined || self.isConnected() === false) {
```
Comment (EnFinlay) [inline lib/client.js]: I'd be happier if this line wasn't repeated.  Maybe a function with one parameter like `repeat` which calls itself after 1 second if `repeat` is false, and errors if `repeat` is true.

**Comment by EnFinlay** [comment]
:+1:

File: lib/client.js (hunk 2 above)
Comment (RongxinZhang) [inline lib/client.js]: Interesting, will try this.

**Pull request merged**